File: tests/src/joda-time/joda-time/2.9.9/src/test/java/joda_time/joda_time/LocalTimeInnerPropertyTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package joda_time.joda_time;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import org.joda.time.LocalTime;
import org.junit.jupiter.api.Test;

public class LocalTimeInnerPropertyTest {

    @Test
    void serializesAndDeserializesSecondOfMinuteProperty() throws Exception {
        LocalTime.Property original = new LocalTime(12, 34, 56, 789).secondOfMinute();

        LocalTime.Property restored = deserialize(serialize(original));

        assertThat(restored.get()).isEqualTo(original.get());
        assertThat(restored.getFieldType()).isEqualTo(original.getFieldType());
        assertThat(restored.getLocalTime()).isEqualTo(original.getLocalTime());
    }

    private static byte[] serialize(Serializable value) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        try (ObjectOutputStream objectOutput = new ObjectOutputStream(output)) {
            objectOutput.writeObject(value);
        }
        return output.toByteArray();
    }

    private static LocalTime.Property deserialize(byte[] serialized) throws IOException, ClassNotFoundException {
        try (ObjectInputStream objectInput = new ObjectInputStream(new ByteArrayInputStream(serialized))) {
            return (LocalTime.Property) objectInput.readObject();
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/LocalTimeInnerPropertyTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

import org.joda.time.DateTimeFieldType
import org.joda.time.LocalTime
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class LocalTimeInnerPropertyTest {
  @Test
  def serializesAndDeserializesLocalTimeProperty(): Unit = {
    val originalTime: LocalTime = new LocalTime(9, 30, 15, 250)
    val originalProperty: LocalTime.Property = originalTime.minuteOfHour()

    val serialized: Array[Byte] = serialize(originalProperty)
    val restoredProperty: LocalTime.Property = deserializeProperty(serialized)

    assertEquals(originalTime, restoredProperty.getLocalTime)
    assertEquals(DateTimeFieldType.minuteOfHour(), restoredProperty.getField.getType)
    assertEquals(30, restoredProperty.get)
  }

  private def serialize(property: LocalTime.Property): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val output: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      output.writeObject(property)
    } finally {
      output.close()
    }
    bytes.toByteArray
  }

  private def deserializeProperty(serialized: Array[Byte]): LocalTime.Property = {
    val input: ObjectInputStream = new ObjectInputStream(new ByteArrayInputStream(serialized))
    try {
      input.readObject().asInstanceOf[LocalTime.Property]
    } finally {
      input.close()
    }
  }
}
